bam_mate.c

```diff
@@ -472,17 +473,486 @@ int bam_sanitize(sam_hdr_t *h, bam1_t *b, int flags) {
+                memcpy((char *)MMq, num, l);
+                MMq += l;
+                *MMq++ = *endptr;
+                n += delta+1;
+                if (ML)
+                    *MLq++ = *MLp++;
+            } else {
+                // next base mod is on boundary of 3' clip point
+                break;
+            }
+
+            MMp = (uint8_t *)endptr;
+            if (*MMp != ',')
+                // error?  if not ; also?
+                break;
+            MMp++;
+        }
+
+        // Copy
+        while (*MMp != ';' && n < counts3[fundamental]) {
+            char *endptr;
+            long delta = strtol((char *)MMp, &endptr, 10);
+            if (counts3[fundamental] - n > delta) {
+                // Copy entire delta in MM and ML including [,;]
+                memmove(MMq, MMp, (uint8_t *)endptr - MMp + 1);
+                MMq += (uint8_t *)endptr - MMp + 1;
+                n += delta+1;
+                if (ML)
+                    *MLq++ = *MLp++;
+            } else {
+                // Next mod is into 3' cutoff, so can terminate MM/ML now
+                n = counts3[fundamental];
+                if (ML)
+                    MLp++;
+            }
+
+            MMp = (uint8_t *)endptr;
+            if (*MMp != ',')
+                break;
+            MMp++;
+        }
+
+        // Skip
+        while (*MMp && *MMp != ';') {
+            while (*MMp && *MMp != ',' && *MMp != ';')
+                MMp++;
+            if (*MMp == ',')
+                MMp++;
+
+            if (ML)
+                MLp++;
+        }
+        MMq[-1] = ';'; // replaces , with ; if clipping right
+        if (*MMp)
+            MMp++;
+    }
+
+    MMp++; // skip nul
+    *MMq++ = 0;
+
+    // Adjust ML B array length
+    if (ML)
+        u32_to_le(MLq-(ML+6), ML+2);
+
+    // Move MM and ML down to include their MM:Z and ML:B bits
+    if (MM) MM-=2;
+    if (ML) ML-=2;
+
+    // Now MM/ML are start of tags, MMq/MLq are ends of edited tags,
+    // and MMp/MLp are ends of original tags.  Walk through tags taking up
+    // any gaps
+    //
+    // Eg XXXXXXmmmmm--YYYlllll-ZZ (m and l are edited MM and ML tags)
+    // => XXXXXXmmmmmYYYlllllZZ
+
+    uint8_t *tag = bam_get_aux(cur), *tag_end = cur->data + cur->l_data;
+    uint8_t *to = tag;
+    while (tag && tag < tag_end) {
+        if (tag[0] == 'M' && (tag[1] == 'M' || tag[1] == 'm')) {
+            // Slow but easy
+            memmove(to, MM, MMq-MM); // length of new tag
+            to += MMq-MM;
+            tag = MMp; // size of old tag
+        } else if (tag[0] == 'M' && (tag[1] == 'L' || tag[1] == 'l')) {
+            memmove(to, ML, MLq-ML);
+            to += MLq-ML;
+            tag = MLp;
+        } else if (tag[0] == 'M' && tag[1] == 'N') {
+            tag = bam_aux_next(cur, tag+2);
+            // Skip it as we'll overwrite this later, although this
+            // does change the tag order.  Instead we could do:
+            *to++ = 'M';
```

Comment:
Pushed a bunch of changes and just noticed some more comments.

I know I had this commented out at one point.  I can't figure out how it resurfaced again, but my edit history is largely gone in squashes now.  Obviously the intention was indeed to not bother filling out MN and the enc function as we always add it at the end anyway.  I think keeping the loop is probably still good for clarity, otherwise it'd be really hard to understand all the pointer manipulation and what it's attempting to do.